### PR TITLE
Raspberry Pi Pico smorgasboard of bugfixes and improvements

### PR DESCRIPTION
**Pull request by ajf58**
### Summary

The purpose of this PR is to get various bugfixes and enhancements found during development of #77368 approved an merged onto `main`.

They are, by design, entirely orthogonal to anything related to the Pico 2, RP2350 SoCs etc.

### Why are there multiple bugfixes in a single PR?

At the extreme, almost every change in this PR could be a one commit : one PR. Creating a single PR has subjective advantages and disadvantages. For the author (me), it's less handle turning. For CI, one PR means less CI compute resource being used. This all works out ok because the project (quite rightly IMHO) doesn't try and automatically squash all commits in a PR.

### Why not wait until #77368 is approved?

- All of the bugfixes are valid for the RP2040 (e.g. `71298136232 drivers: dma: rpi_pico: Correct handling of NULL filter_param`). Getting them onto `main` fixes issues today, decoupling them from new features.
- Some of the changes are entirely cosmetic, e.g. changing whitespace in existing CMake files. They are not necessary for adding new features. When people review #77368 they are concerned about how many files have been touched. Getting these merged in a distinct PR allows the existing PR to be rebased and it'll (hopefully) appear less intimidating.

### CODE REVIEW

File: dts/bindings/clock/raspberrypi,pico-xosc.yaml
Comment (xudongzheng) [inline dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: Since this is set to 1 here, I would suggest removing https://github.com/zephyrproject-rtos/zephyr/blob/0954e443f38775a391aa5fd4bca89159ae6648f9/dts/arm/raspberrypi/rpi_pico/rp2040.dtsi#L165 so the default is consistent across RP2040 and RP2350.

**Review by xudongzheng — CHANGES_REQUESTED** [review]

**Review by tejlmand — CHANGES_REQUESTED** [review]
>  For the author (me), it's less handle turning. 

and for the reviewers ?
So whenever you push changes, do I then have to go through all commits again to see if my comments are now addressed or which other changes you have introduced in the meantime which I might or might not approve ?

While there can be reasons for having multiple commits in one PR, then if those commits are completely orthogonal to each other then please consider opening independent PRs.

Or do not open the PR until you believe all the changes are ready.
Please also take a look at the contribution guideline:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow
> One general practice we encourage, is to make small, controlled changes. This practice simplifies review, makes merging and rebasing easier, and keeps the change history clear and clean.

**Comment by ajf58** [comment]
> and for the reviewers ? 

I can't speak for everyone, but advantages that I can see are that Triage role people don't need to triage 7 PRs, reviewers can (if they wish) see one family of changes.

> So whenever you push changes, do I then have to... <snip/>

Please correct me if I'm wrong, but I don't think that any of the files modified in this draft PR are authored by you, nor are you a maintainer or assignee, so I feel like the very short answer is there's no obligation for you to review this.

> ...go through all commits again to see if my comments are now addressed or which other changes you have introduced in the meantime which I might or might not approve ?

I think yes, this is almost an inevitable side-effect of publicly reviewed PRs. It certainly happens that this situation occurs:
1. Reviewer A requests change 1.
2. Author makes change 1.
3. Reviewer B was happy without change 1, and doesn't like it.
4. Author cannot satisfy the conflicting requirements of Reviewer A and Reviewer B.

I feel that 8 commits pertaining to a single common SoC doesn't impose too much burden on the reviewer.

Regardless of my opinion,  if you disagree and will actively block/NACK this PR when it moves from Draft status, then let's have a concrete proposal, rather than a somewhat open-to-misunderstanding "do it different"

How about if I create 7 (if my maths is correct) PRs (can't add 0954e443f without the preceding commit, [or if I do then the test will run and fail]).

**Comment by tejlmand** [comment]
> Please correct me if I'm wrong, but I don't think that any of the files modified in this draft PR are authored by you, nor are you a maintainer or assignee, so I feel like the very short answer is there's no obligation for you to review this.

I'm maintainer of the build system, and author of HWMv2, so any PR which are working with SoCs and boards might have my interest. Especially PRs which are follow-ups on other PRs I have reviewed.

Now, for this specific PR, I can't tell if one, two, three, or all commits in this PR will have my interest, but more commits added in one PR increase the chance that I have to review the whole PR. If those are individual PRs, then it could be that most of those individual PRs will not have my interest, thus save me overall time.

**Comment by fabiobaltieri** [comment]
> Why are there multiple bugfixes in a single PR?

This is fine, though probably not in your best interest if you want to slipstream the review, the system is going to tag the PR against multiple subsystems, notify all maintainers and probably assign it incorrectly. Then the reviewer see a PR touching multiple things and are naturally inclined to push it forward and wait for someone else to do an initial pass, then wait for the comment to be addressed and do their own part, which is exactly what you see here.

Plus, I mean, the entire PR comment is basically acknowledging that you are trading your time for the reviewers one, how does you expect the other side to react? That conversation does not even belong to here, just leave a summary of the changes and let the reviewers go through them, if the PR spans across multiple subsystems it will need multiple rounds, if it's a new SoC this is inevitable and just part of the process.

Anyway, folks, please let's focus on the change itself and get past the hump.

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by ajf58** [comment]
Thanks for approving the PR, @fabiobaltieri .

I asked if should split this into 7 PRs. I waited... 10 days, got no response, so moved this from draft to review ready. If the maintainers want/mandate that this is split into 7 PRs, please politely and clearly ask me to do so.

> Plus, I mean, the entire PR comment is basically acknowledging that you are trading your time for the reviewers one, how does you expect the other side to react? That conversation does not even belong to here

Thanks, but no, I'll exercise the Right Of Reply.

I expect reviewers to respond as requested [here](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#reviewer-expectations). Sincerely, given that the content of these commits has repeatedly been approved and reviewed as part of another PR, I thought reviewers would think "yes, saving ~70 hours of CI compute time is pragmatic good use of time/compute resource. This is a small bunch of descriptive commits related to a single SoC, some of these commits are literally just whitespace changes, easy peasy, let's get some bugs fixed on `main`. Thanks."

>  just leave a summary of the changes 

https://github.com/zephyrproject-rtos/zephyr/pull/80707/commits I can't think of anything better than the one line summary of each commit message. If the commit messages are confusing/misleading and reviewers have actionable feedback on that, then please do let me know.

**Comment by fabiobaltieri** [comment]
The PR stands up alright in my opinion, just give people time to go over it, we are in hard freeze it won't go until the release is cut anyway.

Don't worry about CI compute time, that's a more systematic problem that needs to be solved at project level. Unless you actively want to work on the more systematic problem of project level CI compute time, then, please, worry about it.

**Review by soburi — DISMISSED** [review]

**Review by kartben — DISMISSED** [review]

**Comment by tejlmand** [comment]
> I asked if should split this into 7 PRs. I waited... 10 days, got no response,

People asked you to do a proper cleanup of your PR in #77368 and you decided to fight that request instead of following the proposed cleanup.

I gave you a bit of advice in https://github.com/zephyrproject-rtos/zephyr/pull/80707#pullrequestreview-2409731787 but what you do with that advice is purely up to you.

I'm not going to tell you what you should or should not do, but i'm going to tell you if I can approve a PR or if I can't and in that case also give you an explanation.

One problem I have when looking at the PR description, is that I cannot judge if more commits will be added later when you find more bug to fix. This again results in an open-ended PR (when will more commits be added ?) and therefore I wait with further review until the PR seems stable (which it appears to be now).

File: CODEOWNERS
Comment (tejlmand) [inline CODEOWNERS:425]: Wrong vendor prefix:
```suggestion
/dts/arm/raspberrypi/rpi_pico/            @yonsch
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: dts/bindings/clock/raspberrypi,pico-xosc.yaml
Comment (tejlmand) [reply dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: @ajf58 can you please address this comment.
Either by following the suggestion or reply why that's not a good idea in this case.

Comment (ajf58) [reply dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: > Either by following the suggestion or reply why that's not a good idea in this case.

@xudongzheng was the original author of this change, so they're kinda arguing with their past self here, with me as a proxy. So the full suite of options are:

1. Leave as I've done. This will not change the behaviour of the in-tree RP2040-based boards. This, implicitly, has been approved many many times over as part of another PR.
2. Only remove `startup-delay-multiplier` from rp2040.dtsi. This will effectively revert what @xudongzheng added in the first place.
3. Remove `startup-delay-multiplier` from rp2040.dtsi and only add it to the in-tree boards where it is known to be needed, i.e. cross match against https://github.com/raspberrypi/pico-sdk/tree/master/src/boards/include/boards . This, IMHO is arguably what the original PR that introduced the delay should have done: it fixes things that are known to need this, and doesn't alter out-of-tree boards' behaviour.

**Comment by soburi** [comment]
Just a comment,
Regarding the PR split, I think it's fine as is.

This has already been explored in detail in #77368, and is in fact a collection of parts that can be rebased and extracted without being related to rp2350.

**Comment by tejlmand** [comment]
> Regarding the PR split, I think it's fine as is.

@soburi now is it ?

@ajf58 is now having two PRs opened, this one with 8 commits and #77368 with 28 commits, 8 of those being the same commits in both PRs.

I cannot judge whether I should put my comments in this PR or in the other PR.
There is no squash / mentioning if one PR is dependent on the other, or if the should be merged in a specific order.
On top of that, the PR description was open-ended.

I currently have on github (and there are many people in Zephyr assigned even more PRs):
26 PRs where i'm assignee
217  PRs where i'm requested as reviewer.

Properly crafted PRs and logical splitted PRs makes reviewers life so much easier.
Not to mention proper PR descriptions are extremely important to make reviewing of PRs as efficient as possible.

This is what I tried to explain to @ajf58 in my comment: https://github.com/zephyrproject-rtos/zephyr/pull/80707#pullrequestreview-2409731787

An example of two commits, where one is a cleanup, and the second depends on the first.:
https://github.com/zephyrproject-rtos/zephyr/pull/41301
https://github.com/zephyrproject-rtos/zephyr/pull/40555
and where #40555 originally included #41301 as a single squashed commit with commit message like this:
```
squashed commit of #41301, not to be reviewed
```
So that once #41301 got merged, then #40555 could be rebased and the `not to review` commit was gone.

Pros:
- makes it very clear which code changes should be reviewed in which PR.
- makes it very clear the order of which those PRs should be merged.
- Have very clear PR description, and not like `The purpose of this PR is to get various bugfixes and enhancements found during development of https://github.com/zephyrproject-rtos/zephyr/pull/77368 approved an merged onto main.` which tells me nothing about when this PR is ready to be reviewed.
 
You may say that this PR as it stands today, makes it ready for review, and thus has the potential for approval, only outstanding comment is the CODEOWNERS comment, and that has only been addressed in #77368.

But at the time I made the https://github.com/zephyrproject-rtos/zephyr/pull/80707#pullrequestreview-2409731787 comment, then the direction of this PR was not clear at all.

File: dts/bindings/clock/raspberrypi,pico-xosc.yaml
Comment (tejlmand) [reply dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: > @xudongzheng was the original author of this change, so they're kinda arguing with their past self here,

Ever written code where you later discover this should probably have been done differently or at another location ?
I have done that many times.

Of course I cannot answer for @xudongzheng , but to me this looks like such a case, and when this PR broadly states `improvements`, then this seems to be one such improvement which can be taken in.

Though I agree it more seems like something which should be a regular comment, and not a change request.

> Remove startup-delay-multiplier from rp2040.dtsi and only add it to the in-tree boards where it is known to be needed, i.e. cross match against https://github.com/raspberrypi/pico-sdk/tree/master/src/boards/include/boards . This, IMHO is arguably what the original PR that introduced the delay should have done: it fixes things that are known to need this, and doesn't alter out-of-tree boards' behaviour.

To me this also sounds like the ideal approach.

And as PR author you are free to decide if you want to include this change or not.

But thanks for replying to the comment, as that is the only way for others to know if it has been seen / considered.

**Comment by soburi** [comment]
@tejlmand 

I understand the rationality of what you say, but since it is a "recommendation" in the guidelines, I do not think it is completely legitimate.

Regarding the background of this PR, I think that it is an excerpt from #77368 where the discussion has already converged to a certain extent, and that should be taken into consideration. I think that the point you pointed out should be pointed out as a carelessness on the part of reviewers, including myself.

It is true that there may be some parts that are not explained well, but it is impossible for those outside of the inner circle (which can be rephrased as full-time contributors belonging to the sponsoring company of this project) to understand the degree of the binding force of the "recommendation" in the guidelines. If that is the case, the guideline should be revised accordingly, and I think that @ajf58's current response is sufficiently rational.
Just as a reviewer's time is finite, so is a contributor's time.

**Comment by tejlmand** [comment]
> but since it is a "recommendation" in the guidelines, I do not think it is completely legitimate.

True it's only recommendation, which is also the reason i'm not blocking PRs on this ground, and i'm not blocking #77368.
Others might, but I haven't.

Reason I blocked this PR when it was still draft was due to how it was open-ended, and I wanted to avoid a situation where new changes are added later without proper review. It sometimes happens that when enough reviewers has approved then the PR gets merged and I wanted to avoid that situation for this PR, due to above.

> I think that it is an excerpt from https://github.com/zephyrproject-rtos/zephyr/pull/77368 where the discussion has already converged to a certain extent.

This PR description just states:
> > various bugfixes and enhancements found during development of https://github.com/zephyrproject-rtos/zephyr/pull/77368

It doesn't state whether content of this PR is already discussed elsewhere. (@soburi as assignee of the other PR you probably have followed the discussion in #77368 more closely than me)
There are 352 comments by now in #77368, and i haven't read each and everyone of these.

> but it is impossible for those outside of the inner circle ... to understand the degree of the binding force of the "recommendation" in the guidelines.

True, it's only recommendations.
But as reviewer it's still important to point to those recommendations and and advice / guide contributors.
Especially I want to challenge the author when he writes:
> >  For the author (me), it's less handle turning. 

Minimizing the time spent overall by n-number of persons is imo more important then the time spent of a single individual.

This is also in line with the [Contributor Expectations](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations):
> > Smaller pull requests (PRs) have the following benefits:
> >
> > - Reviewed more quickly and reviewed more thoroughly. It’s easier for reviewers to set aside a few minutes to review smaller changes several times than it is to allocate large blocks of time to review a large PR.

-------

Currently this PR is not nack'ed because of its structure or discussions like this comment but because https://github.com/zephyrproject-rtos/zephyr/pull/80707#discussion_r1851001011 haven't been addressed.

File: CODEOWNERS
Comment (tejlmand) [reply CODEOWNERS:425]: > I would just drop this change altogether. See #82230

Sure, if #82230 gets merged first (which I guess it will).

But until then, if this PR contains a change to `CODEOWNERS` then such change should be correct.

Comment (kartben) [reply CODEOWNERS:425]: My point was that the CODEOWNERS file is plain stale anyway and full of incorrect entries (it wasn't even covered with hwmv2 migration and nobody noticed/bothered... https://github.com/zephyrproject-rtos/zephyr/commit/8dc3f856229ce083c956aa301c31a23e65bd8cd8#diff-fcf14c4b7b34fe7a11916195871ae66a59be87a395f28db73e345ebdc828085b). Dropping this change is effectively a simple way to avoid a possible future merge conflict IMO. 

BTW if the change stays (which I really recommend it doesn't), why is the code owner not being updated to @soburi?

Comment (ajf58) [reply CODEOWNERS:425]: > Dropping this change is effectively a simple way to avoid a possible future merge conflict IMO.

Agreed. We should just wait the ~48 hours for #82230 . Absolute worse case scenario would be a known-to-be-wrong file is made a tiny bit more wrong, before being deleted.

> if the change stays (which I really recommend it doesn't), why is the code owner not being updated to @soburi?

My assumption was the file was correct at the point I changed it, which it turns out was not the case, as you've pointed out.

File: dts/bindings/clock/raspberrypi,pico-xosc.yaml
Comment (ajf58) [reply dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: > you are free to decide if you want to include this change or not.

I'm not going to be extending the scope of this change beyond what it currently does, i.e. "prevent out-of-tree boards based that use this DTS binding requiring to define a parameter." 

It's a remarkable coincidence that in the "future" (subsequent commits)[1] it means that if another SoC uses this same driver it won't have to keep defining something to be equal to one. Someone should make one of those and put it on some boards... maybe get a PR to put them in-tree.

The value is board, not SoC specific, and really shouldn't be in rp2040.dtsi whatsoever. @xudongzheng , perhaps you can raise an issue referencing back to your own PR with these suggestions? I don't have time to do that, nor to action it.

[1] or ~weeks in the past for those humans who measure their life like this.

Comment (soburi) [reply dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: Adding a default value on the yaml side will not change the value explicitly specified in dts.
In other words, the current behavior will not change even if this change is.
Therefore, I think it would be better to extract this into a separate PR.

As @xudongzheng says, deleting the value on the dtsi side would be better.
Can you please address this after this PR is completed? > @xudongzheng

Comment (xudongzheng) [reply dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: I have it in my internal RP2350 notes to follow up on `startup-delay-multiplier` with a PR if it's not resolved here.

The important part of #78200 is making the delay configurable, not so much what the default value is. I prefer 64, @soburi somewhere in #77368 mentioned preferring 1 to be consistent with the Pico SDK. I'm inclined to defer to his opinion as the maintainer.

**Comment by ajf58** [comment]
My plan of action:

1. Let's get some green ticks against this.
2. Wait for #82230 to be merged (I think it's just running down the clock on https://merge-list.zephyrproject.io/)
3. Merge this.
4. I'll rebase other PRs on top of it.

**Comment by soburi** [comment]
@ajf58 
#82230 has merged.

File: CODEOWNERS
Comment (ajf58) [reply CODEOWNERS]: This was a blocking request for change, but now it's overtaken by events as the underlying file has been deleted.

I hope this conversation can now be resolved, and we get some approvals against this.

**Comment by ajf58** [comment]
@soburi 
> #82230 has merged.

Indeed it has, making the previous request for change moot. I've re-requested review.

**Review by soburi — APPROVED** [review]
LGTM

File: dts/bindings/clock/raspberrypi,pico-xosc.yaml
Comment (soburi) [reply dts/bindings/clock/raspberrypi,pico-xosc.yaml:14]: @xudongzheng 
The dts setting takes precedence over the default, so even if we do not fix it in this PR, the behavior will not change. Let's deal with it in another PR.

**Review by tejlmand — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Pull request merged**
